### DISCUSSION TITLE
Building without NodeJS/WASM

### DISCUSSION THREAD

**Discussion by fruffy** [p4lang/p4analyzer#27] [Q&A] [answered]
I would like to try out the language server with my Sublime/Vim-LSP setup instead of VSCode. Is it possible to build the server executable using only Rust's `cargo`? Or is the NodeJS/WASM setup always required?

**Comment by AndrewF001** [comment, marked as answer]
NodeJS isn't a requirement if you're not interested in the `.vsix` file, as the Rust competent can be compiled and run independently. The WASM is a separate cargo, so to avoid building it, you can run this command: `cargo build --bin p4analyzer --release`, and the executable will be created at `target\release\p4analyzer.exe`.

We haven't tested how the program interacts with other editors, but it uses the LSP 1.17 standard of communication through stdin & stdout. Please keep in mind that this project is very early stages, so not many features have been implemented, but let me know if you have any issues compiling the project!

**Reply by fruffy** [reply]
Great, thanks for the prompt reply!

**Reply by fruffy** [reply]
To install the binary in the cargo folder I had to run `cargo install --bin p4analyzer --path crates/p4-analyzer/`. Is this the right/intended command?

**Reply by AndrewF001** [reply]
I believe [cargo install](https://doc.rust-lang.org/cargo/commands/cargo-install.html#description) will only add it to the Create install list which is used for the Rust Cargo package manager. `Cargo build` should create a [self-contained executable](https://stackoverflow.com/a/60947511) which can be called by Sublime/Vim-LSP. I'm a bit out of my depths but something like this should work for [Sublime](https://lsp.sublimetext.io/client_configuration/)?:
Open Preferences > Package Settings > LSP > Settings and add "P4Analyzer" client configuration to the "clients": 
``` bash
{
    "clients": {
        "p4-analyzer": {
             // enable this configuration
             "enabled": true,
             // the startup command -- what you would type in a terminal
             "command": ["PATH/TO/p4analyzer", "Create-a-CLI-command"],
             // the selector that selects which type of buffers this language server attaches to
            "selector": "source.p4",
        }
    }
}
```

**Reply by fruffy** [reply]
Oh, I was just wondering whether you had a standard way to install the binary generated by cargo build in an install folder. For example `~./local/bin`, `/usr/local/bin`, or `~/.cargo/bin`. The command I used seems to work. 

Yes, the sublime config should work. If I have some time, I will try to create a package for it.

**Reply by AndrewF001** [reply]
Glad it worked for you! If you have the time, let me know how it goes. I have created an issue based on this discussion (#28) if you want to contribute!
